### PR TITLE
Progress with testing arbitrator

### PR DESCRIPTION
**Pull request by HamishBrownPFR**
Working on #6763

### CODE REVIEW

**Comment by HamishBrownPFR** [comment]
@hol353 @par456 can ypu please review and merge if happy.  Code changes to the new arbitrator so should not be affecting anything in release

File: Models/PMF/Biomass/Biomass.cs [hunk 1]
```diff
@@ -203,6 +203,20 @@ public Biomass(Biomass from)
             _MetabolicN = from.MetabolicN;
         }
 
+        /// <summary>Initializes a new instance of the <see cref="Biomass"/> class from the OrganNutrientState passed in</summary>
+        /// <param name="ons">ons.</param>
+        public Biomass(OrganNutrientsState ons)
```
Comment (hol353) [inline Models/PMF/Biomass/Biomass.cs:208]: By including this here, the Biomass class knows about OrganNutirentState. This doesn't feel right. Instead, have a method in OrganNutirentState called something like ToBiomass that creates an Biomass instance.

File: Models/PMF/BiomassArbitration/BiomassArbitrator.cs [hunk 2]
```diff
@@ -78,6 +84,53 @@ public class BiomassArbitrator : Model
         [Link(Type = LinkType.Child, ByName = true)]
         public PlantNutrientsDelta Nitrogen { get; private set; }
 
+        /// <summary> The amount of carbon fixed by photosynthesis</summary>
+        public double TotalCFixationSupply { get { return Carbon.TotalFixationSupply; } }
+
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/BiomassArbitrator.cs:89]: These sorts of property gets can be simplified like this:
```c#
public double TotalCFixationSupply => Carbon.TotalFixationSupply;
```

File: Models/PMF/BiomassArbitration/BiomassArbitrator.cs [hunk 3]
```diff
@@ -198,13 +257,63 @@ virtual protected void OnDoActualPlantPartioning(object sender, EventArgs e)
...
+                foreach (Organ o in PlantOrgans)
+                {
+                    checkC += (o.Carbon.SuppliesAllocated.Fixation + o.Carbon.SuppliesAllocated.Uptake - o.Detached.C - o.LiveRemoved.C - o.DeadRemoved.C);
+                    checkN += (o.Nitrogen.SuppliesAllocated.Fixation + o.Nitrogen.SuppliesAllocated.Uptake - o.Detached.N - o.LiveRemoved.N - o.DeadRemoved.N);
+                    endC += o.C;
+                    endN += o.N;
+                }
+
+                double cdiff = Math.Abs(checkC - endC);
+                if (cdiff > 1e-12)
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/BiomassArbitrator.cs:304]: Don't we have a MathUtilities function that does this?

File: Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs [hunk 4]
```diff
@@ -22,8 +25,13 @@ public class DeficitDemandFunction : Model, IFunction
         public IFunction multiplier = null;
 
         private Organ parentOrgan = null;
-        private OrganNutrientDelta dCarbon = null;
-        private OrganNutrientDelta dNitrogen = null;
+        private OrganNutrientDelta dNutrient = null;
+        //private OrganNutrientDelta dCarbon = null;
+        //private OrganNutrientDelta dNitrogen = null;
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs:30]: Why is the variable called dNutrient? What does the 'd' stand for? Also remove commented code.

File: Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs [hunk 5]
```diff
@@ -33,16 +41,29 @@ private Organ FindParentOrgan(IModel model)
...
+                throw new Exception(Name + "cannot find parent Nutrient Delta");
+
+            return FindParentNutrientDelta(model.Parent);
+        }
+
+        private string FindPoolName(IModel model)
+        {
+            if ((model.Name == "Structural")|| (model.Name == "Metabolic")|| (model.Name == "Storage"))
+                return model.Name;
+            if ((model is OrganNutrientDelta)||(model is Organ) || (model is IPlant) || (model is IZone) || (model is Simulation))
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs:63]: Yuk. This looks terrible. Is there a better way to find a pool? I'm not sure how this is supposed to work but do pools have an interface that you can look for with a FindInScope call?

File: Models/PMF/BiomassArbitration/Functions/MobilisationSupplyFunction.cs [hunk 6]
```diff
@@ -86,7 +86,7 @@ private double CalcAmountToReallocate(double sourceAmount, List<IFunction> child
             multiplier = Math.Max(0, Math.Min(1, multiplier));
 
             //sourceAmount ie: Leaf.Carbon.Live.Metabolic
-            double reAllocationsupply = sourceAmount * multiplier * parentOrgan.senescenceRate;
+            double reAllocationsupply = sourceAmount * multiplier * parentOrgan.theSenescenceRate;
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/Functions/MobilisationSupplyFunction.cs:89]: Remove the 'the' prefix.

File: Models/PMF/BiomassArbitration/RootNetwork.cs [hunk 7]
```diff
@@ -201,7 +202,10 @@ public double[] LengthDensity
         /// <summary>Root depth.</summary>
         [JsonIgnore]
         [Units("mm")]
-        public double Depth { get { return PlantZone.Depth; } }
+        public double Depth { 
+            get { return PlantZone.Depth; }
+            set { PlantZone.Depth = value; }
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/RootNetwork.cs:207]: Is a setter needed? What would call it?

File: Models/PMF/BiomassArbitration/Structures/IParentOfNutrientsPoolState.cs [hunk 8]
```diff
@@ -5,8 +5,7 @@
     /// </summary>
     public interface IParentOfNutrientsPoolState
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/Structures/IParentOfNutrientsPoolState.cs:6]: Is this still needed?

File: Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs [hunk 9]
```diff
@@ -40,31 +40,86 @@ public NutrientsStates(double c, double n, double p, double k)
     public class OrganNutrientsState : Model, IParentOfNutrientsPoolState
     {
         /// <summary> The weight of the organ</summary>
-        public NutrientPoolsState Weight { get; private set; }
+        public NutrientPoolsState Weight
+        {
+            get
+            {
+                return Cconc > 0 ? Carbon / Cconc : new NutrientPoolsState(0, 0, 0);
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs:47]: These can all be simplified as per earlier comment.

File: Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs [hunk 10]
```diff
@@ -143,79 +181,79 @@ public OrganNutrientsState()
             Phosphorus = new NutrientPoolsState();
             Potassium = new NutrientPoolsState();
             Cconc = 1.0;
-            UpdateProperties();
         }
 
         /// <summary>return pools divied by value</summary>
-        public static OrganNutrientsState operator /(OrganNutrientsState a, double b)
+        public static OrganNutrientsState divide (OrganNutrientsState a, double b, double cconc)
```
Comment (hol353) [inline Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs:187]: These should be upper case.

**Review by hol353 — CHANGES_REQUESTED** [review]

File: Models/PMF/Biomass/Biomass.cs (hunk 1 above)
Comment (HamishBrownPFR) [reply Models/PMF/Biomass/Biomass.cs:208]: Change made as suggested.

File: Models/PMF/BiomassArbitration/BiomassArbitrator.cs (hunk 2 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/BiomassArbitrator.cs:89]: Change made

File: Models/PMF/BiomassArbitration/BiomassArbitrator.cs (hunk 3 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/BiomassArbitrator.cs:304]: Yes, FloatsAreEquat.  Thanks

File: Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs (hunk 4 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs:30]: stands for delta.  Have changed variable name to organNutrientDelta to be clearer

File: Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs (hunk 5 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/Functions/DeficitDemandFunction.cs:63]: This was more complex than needed.  Have reduced it to a name check to throw an error if name is not in required list.

File: Models/PMF/BiomassArbitration/Functions/MobilisationSupplyFunction.cs (hunk 6 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/Functions/MobilisationSupplyFunction.cs:89]: Done

File: Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs (hunk 9 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs:47]: done

File: Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs (hunk 10 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/Structures/OrganNutrientsState.cs:187]: Fixed.  I wonder why they were working?

File: Models/PMF/BiomassArbitration/Structures/IParentOfNutrientsPoolState.cs (hunk 8 above)
Comment (HamishBrownPFR) [reply Models/PMF/BiomassArbitration/Structures/IParentOfNutrientsPoolState.cs:6]: No, Removed

**Comment by HamishBrownPFR** [comment]
@hol353 requested changes made.  Thanks

**Review by hol353 — APPROVED** [review]

**Pull request merged**